data-pipeline/src/telemetry/metrics.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+//! Provides an abstraction layer to hold metrics that comes from 'SendDataResult'.
+use ddcommon::tag;
+use ddtelemetry::data::metrics::{MetricNamespace, MetricType};
+use ddtelemetry::metrics::ContextKey;
+use ddtelemetry::worker::TelemetryWorkerHandle;
+use std::collections::HashMap;
+
+/// trace_api.requests metric
+pub const API_REQUEST_STR: &str = "trace_api.requests";
+/// trace_api.errors metric
+pub const API_ERRORS_STR: &str = "trace_api.errors";
+/// trace_api.bytes metric
+pub const API_BYTES_STR: &str = "trace_api.bytes";
+/// trace_api.responses metric
+pub const API_RESPONSES_STR: &str = "trace_api.responses";
+/// trace_chunk_sent metric
+pub const CHUNKS_SENT_STR: &str = "trace_chunk_sent";
+/// trace_chunk_dropped metric
+pub const CHUNKS_DROPPED_STR: &str = "trace_chunk_dropped";
+
+struct Metric {
+    name: &'static str,
+    metric_type: MetricType,
+    namespace: MetricNamespace,
+}
+
+const METRICS: &[Metric] = &[
+    Metric {
+        name: API_REQUEST_STR,
+        metric_type: MetricType::Count,
+        namespace: MetricNamespace::Tracers,
+    },
+    Metric {
+        name: API_ERRORS_STR,
+        metric_type: MetricType::Count,
+        namespace: MetricNamespace::Tracers,
+    },
+    Metric {
+        name: API_BYTES_STR,
+        metric_type: MetricType::Distribution,
+        namespace: MetricNamespace::Tracers,
+    },
+    Metric {
+        name: API_RESPONSES_STR,
+        metric_type: MetricType::Count,
+        namespace: MetricNamespace::Tracers,
+    },
+    Metric {
+        name: CHUNKS_SENT_STR,
+        metric_type: MetricType::Count,
+        namespace: MetricNamespace::Tracers,
+    },
+    Metric {
+        name: CHUNKS_DROPPED_STR,
+        metric_type: MetricType::Count,
+        namespace: MetricNamespace::Tracers,
+    },
+];
+
+/// Structure to accumulate partial results coming from sending traces to the agent.
+#[derive(Debug)]
+pub struct Metrics(HashMap<String, ContextKey>);
```

Comment:
I'm not a fan of using hashmaps to lookup harcoded keys, for performance and correctness,  but I don't really have a good counter proposition that'd be shorter to write...
Maybe associate an index to each metric name and hold Contexts in a Vec/array?

Reply:
I thought about that when coding it but chose the easy path. Since the input is known ahead of time we could save some time from the hashing function. I would try some implementation and run some benchmarks.

Reply:
No objection to starting off with a simpler implementation. But would it be a big deal to add a benchmark test to measure this? It'll be easier for us to observe performance changes in the future.

Reply:
Yeah, of course, it shouldn't be that expensive

Reply:
I did it anyway. When hardcoding the tags for the `trace_api.error` metrics I realized I could squeeze it without a major refactor.